aimmo-game/tests/functional/test_effect_expiry.py

```diff
@@ -0,0 +1,169 @@
+from unittest import TestCase
+
+from mock_world import MockWorld
+
+from simulation.location import Location
+from simulation.pickups import DamageBoostPickup, InvulnerabilityPickup
+
+
+class TestEffectExpiry(TestCase):
+    """
+    Tests if effects of pickups can expire correctly in multiple scenarios. Note: applies only to pickups that inherit
+    TimedEffect!
+    """
+
+    def setUp(self):
+        """
+        Mock a game for each test individually. MockWorld() will set up a:
+        avatar manager, game state, turn manager and a map generator.
+        """
+        SETTINGS = {
+            'START_HEIGHT': 5,
+            'START_WIDTH': 15,
+            'OBSTACLE_RATIO': 0,
+        }
+
+        self.game = MockWorld(SETTINGS)
+        self.game.game_state.add_avatar(1, None, Location(0, 0))
+        self.avatar = self.game.avatar_manager.get_avatar(1)
+        self.cell = self.game.game_state.world_map.get_cell(Location(1, 0))
+
+    def test_single_damage_boost_pickup_expiry(self):
+        """
+        Avatar spawns at ORIGIN. DamageBoostPickup is at 1,0. Avatar moves to the pickup and picks it up next turn and
+        then we wait for the effect to expire EFFECT_TIME turns later (value defined in the effects class).
+        """
+        pickup_created = DamageBoostPickup(self.cell, 10)
+        self.cell.pickup = pickup_created
+        self.assertEqual(self.avatar.attack_strength, 1)
+
+        # Avatar moves EAST to (1,0) where pickup is located, then repeats it 5 times.
+        for i in range(6):
+            self.game.turn_manager._run_single_turn()
+
+        self.assertTrue(isinstance(list(self.avatar.effects)[0], pickup_created.EFFECT))
+        self.assertEqual(list(self.avatar.effects)[0]._time_remaining, 5)
+        self.assertEqual(self.avatar.attack_strength, 11)
+
+        # Run 5 more turns and expect the effect to expire.
+        for i in range(5):
+            self.game.turn_manager._run_single_turn()
+
+        self.assertEqual(len(self.avatar.effects), 0)
+        self.assertEqual(self.avatar.attack_strength, 1)
+
+    def test_single_invulnerability_pickup_pickup_expiry(self):
+        """
+        Avatar spawns at ORIGIN. InvulnerabilityPickup is at 1,0. Avatar moves to the pickup and picks it up next turn
+        and then we wait for the effect to expire EFFECT_TIME turns later (value defined in the effects class).
+        """
+        pickup_created = InvulnerabilityPickup(self.cell)
+        self.cell.pickup = pickup_created
+        self.assertEqual(self.avatar.resistance, 0)
+
+        # Avatar moves EAST to (1,0) where pickup is located, then repeats it 5 times.
+        for i in range(6):
+            self.game.turn_manager._run_single_turn()
+
+        self.assertTrue(isinstance(list(self.avatar.effects)[0], pickup_created.EFFECT))
+        self.assertEqual(list(self.avatar.effects)[0]._time_remaining, 5)
+        self.assertEqual(self.avatar.resistance, 1000)
```

Comment:
where does the 1000 come from?

Reply:
That's the resistance effect that the InvulnerabilityPickupEffect causes the avatar to have.

Reply:
Yes, but shouldn't that be parametrised somewhere too?